File: packages/schemas/src/recipe.ts
import { z } from 'zod';

import { Timestamps } from './shared';

export const Recipe = z
  .object({
    id: z.uuid(),
    bookId: z.uuid(),
    name: z.string().min(1),
    pageStart: z.number().int().positive().nullable(),
    pageEnd: z.number().int().positive().nullable(),
  })
  .extend(Timestamps.shape);

export const CreateRecipe = z.object({
  bookId: z.uuid(),
  name: z.string().min(1),
  pageStart: z.number().int().positive().nullable().optional(),
  pageEnd: z.number().int().positive().nullable().optional(),
});

export const SearchQuery = z.object({
  q: z.string().min(1),
});

export type RecipeType = z.infer<typeof Recipe>;
export type CreateRecipeInput = z.infer<typeof CreateRecipe>;

File: apps/ios/Slowcook/Models/Recipe.swift
import Foundation

struct Recipe: Identifiable {
    let id: String
    let bookId: String
    var name: String
    var pageStart: Int?
    var pageEnd: Int?
    let createdAt: Date

    init(id: String = UUID().uuidString, bookId: String, name: String, pageStart: Int?, pageEnd: Int?, createdAt: Date = Date()) {
        self.id = id
        self.bookId = bookId
        self.name = name
        self.pageStart = pageStart
        self.pageEnd = pageEnd
        self.createdAt = createdAt
    }

    var pageDisplay: String {
        if let start = pageStart, let end = pageEnd, end != start {
            return "pp. \(start)–\(end)"
        } else if let start = pageStart {
            return "p. \(start)"
        }
        return ""
    }
}
